src/Interpreters/Context.cpp

```diff
@@ -3235,10 +3235,12 @@ MarkCachePtr Context::getMarkCache() const
 
 void Context::clearMarkCache() const
 {
-    std::lock_guard lock(shared->mutex);
+    /// Get local shared pointer to the cache
+    MarkCachePtr cache = getMarkCache();
 
-    if (shared->mark_cache)
-        shared->mark_cache->clear();
```

Comment:
Maybe add this change for other `clear*Cache` methods (primary index, uncompressed, ...)?